File: core/java/src/main/java/hara/lang/protocol/IDeref.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.ideref", name = "IDeref")
public interface IDeref<V> {
  @HaraMethod(value = "deref", arity = 1)
  V deref();
}

File: core/rust/src/lang/protocol/ideref.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.ideref", name = "IDeref")]
pub trait IDeref {
    type Output;

    #[hara_method(value = "deref", arity = 1)]
    fn deref(&self) -> Self::Output;
}
